src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioDocumentNavigationService.cs

```diff
@@ -176,12 +176,10 @@ public bool TryNavigateToPosition(Workspace workspace, DocumentId documentId, in
             return NavigateTo(textBuffer, vsTextSpan);
         }
 
-        private static Document OpenDocument(Workspace workspace, DocumentId documentId, bool usePreviewTab)
+        private static Document OpenDocument(Workspace workspace, DocumentId documentId, bool reopenDocument, bool usePreviewTab)
         {
-            // Always open the document again, even if the document is already open in the 
-            // workspace. If a document is already open in a preview tab and it is opened again 
-            // in a permanent tab, this allows the document to transition to the new state.
-            if (workspace.CanOpenDocuments)
+            if (workspace.CanOpenDocuments &&
+                (reopenDocument || !workspace.IsDocumentOpen(documentId)))
```

Comment:
Right, we call Workspace.OpenDocument again, but now only for NavigateTo scenarios. A host could reject OpenDocument calls on documents that are already open, but the VisualStudioWorkspaceImpl does handle it. 

NavigateTo just tells us to navigate to the chosen item both when the item is previewed and when it's chosen, with no flag indicating which is occurring. They set the new document state scope before calling us, and we have to re-open the document for that state to be observed.

Are you suggesting that NavigateTo should call into a separate API that knows how to re-open documents and remove that ability from the VisualStudioWorkspaceImpl, or just that we should formally document that hosts must handle requests to open already-open documents as they see fit?